### DISCUSSION TITLE
"Ain't broke, don't fix it" or "REFACTOR"?

### DISCUSSION THREAD

**Discussion by limefrogyank** [BlazorFluentUI/BlazorFluentUI#387] [General]
I just took a look at the Microsoft FluentUI React repo.  It's VERY different now.  It seems they've been refactoring all their code to mimic webcomponent styles.  I haven't followed React programming for a while now, but maybe React has gone the same way.  Basically, components have slots and you fill slots with things...  (very simplified explanation)

I was trying to add the new icons to `Checkbox` and wanted to see what styling updates I needed to make... but it's just too different.  So I'll either have to just make it work... or refactor the entire thing to match what they have.  

How much do we want to keep things the same?  It's probably not worth it much at this point.  But I don't know how much they're going to change styling in the future.

**Comment by limefrogyank** [comment]
Ah, color names are VERY different now, too.

**Comment by GurliGebis** [comment]
I think following Microsoft would be the best way to go forward.
If not, then it will end up being too different, and difficult for people to switch to blazor if they have used the react version before.

**Comment by vnbaaij** [comment]
Eventually the stying will be aligned to the webcomponents way. This is now happening with fluent implementation of the FAST stack. 

Microsoft is working an a Blazor 'wrapper' (https://github.com/microsoft/fast-blazor) for these fluent webcomponents. Is has been brought up by @limefrogyank before, is not there yet but I'm expecting it to mature and grow quickly. Contributions are very welcome as the team is not coming from a Blazor background (but from the webcomponent side). They are being suported by the .NET Blazor team as well. I already contributed to the library and hope others will as well. That way we can leverage some of the learnings from BlazorFluentUI.

Refactoring will be a boatload of work and we will probably always be a couple of step behind the work Microsoft is doing. Once Microsoft will release a V1 of this library, I think significantly more people will choose such an offical library over our 'hobby project' (not meant derrogative but as in people building it in their spare time)

I'm more on the "aint't broke, don't fix it" side. We had a good run with this but maybe it is time to join forces and work together on the offical library from Microsoft.

**Reply by limefrogyank** [reply]
Yes, I agree with this.

**Reply by gallwood** [reply]
I'm not a contributor to this repo but I am invested in it. I looked at the FAST blazor implementation mentioned above but (not having much time) didn't fully understand how mature this is. 

The nice thing about BlazorFluentUI is that is seems almost complete and mirrors the react components pretty well.

Do you know how far off they are with having feature parity with this project?

Maybe it doesn't work the same, i.e. there is no special implementation specifically, all it does is wrap the JS library?

Sorry if this is a confusing post. It mimics my mind right now :)

**Reply by limefrogyank** [reply]
None of the more complex components are there.  What they have are basic buttons, cards, sliders, etc.  No calendar component or virtualized list.  They are completely different from the react components.  Instead, they are built using web components.  It's a newer HTML thing that uses something called a shadow DOM.  You're actually creating custom html elements when you create a "web component".  FAST is just a way of implementing a web component that makes them easier to use and modify.  One nice thing about them is that they're super responsive.  But since you're making custom html elements, they take text attributes as parameters.  You'll have to use javascript to make them accept anything more complex.

FAST Blazor is just a Blazor wrapper around the web components made using FAST.

**Reply by gallwood** [reply]
Aha, thanks for the explanation.

Will some of the more complex components (Date Picker, virualised lists) be something that will eventually added, do you think?

**Reply by limefrogyank** [reply]
Not sure how up-to-date this is, but these are the regular FAST components.  You can get a "Fluent" version of these from the fluent repo that look more like what you'd expect.

https://explore.fast.design/components/

**Comment by vnbaaij** [comment]
To add to what Lee just said, the complex components are being worked on. For example, I have added a wrapper for the Fast Data Grid component (bit like Details List). Not as complete yet, but we'll get there. DatePicker is in the works.
It is just impossible for the handful of people we are to keep up with what is being done at the React side. That is why I am shifting my attention to contributing to the Microsoft Fast/Fluent Blazor wrapper

**Comment by vnbaaij** [comment]
This is the most up to date overview of the fluent web components: https://fluent-components.azurewebsites.net/?path=/story/accordion--base

________________________________

**Comment by gallwood** [comment]
Thanks @vnbaaij  for the link. So from what I can see and understand, the fluent web components are at the moment a core set of controls (I don't know if they should be called controls :) ). 

From the repo for this, it appears there are 3 different packages. React; React NorthStart and Web Components. The first two are used today in Office and MS Teams. The later in Edge. 

Using the Fluent UI React site [https://developer.microsoft.com/en-us/fluentui#/controls/web/button](https://developer.microsoft.com/en-us/fluentui#/controls/web/button) as an example, it seems to offer more than the fluent web components, for example split buttons, CommandBarButtons, Nav etc. 

As I'm not a UX  expert, I kind of relied upon libraries like BlazorFluentUI to help me achieve a nice looking, functional app.

My biggest question right now is how I get from web components to something like BlazorFLuentUI if work on this project stops. Where will I get SplitButtons from, what about Models and Panels?

Is the aim of the FAST web components to supersede  what you and other have already achieved in BlazorFluentUI?

**Reply by vnbaaij** [reply]
> Thanks @vnbaaij for the link. So from what I can see and understand, the fluent web components are at the moment a core set of controls (I don't know if they should be called controls :) ). 

They are not. They are called components :)
 
> From the repo for this, it appears there are 3 different packages. React; React NorthStart and Web Components. The first two are used today in Office and MS Teams. The later in Edge.

Correct. You should translate this to React == production, Northstar == React.vNext and WC == generic Fluent UI implementation **not** based on React (but on a implementation of the FAST framework) and useable from a lot of (frontend) frameworks like Angular, Vue, React, ASP.NET and Blazor.

> Using the Fluent UI React site https://developer.microsoft.com/en-us/fluentui#/controls/web/button as an example, it seems to offer more than the fluent web components, for example split buttons, CommandBarButtons, Nav etc.

Yes, but you should not compare the different libraries like that. Fluent UI does not describe the controls and/or how they work. See it more as a set of design guidelines. The plan of the FAST team is to extend the Fluent UI WC set of components so it will offer a rich set of components and functionalities. Not to exactly replicate the functionality of the React libraries.
 
> As I'm not a UX expert, I kind of relied upon libraries like BlazorFluentUI to help me achieve a nice looking, functional app.

And you still can. BlazorFluentUI is not going away. We are just not going to add new functionalities to it and consider the version that is out now as 'done'. That will most probably also mean that we won't be doing a lot of bug fixing anymore. Only thing we intend to do is get a V6 version out of the door when .NET 6 is released (on Nov 9th). 
As said before, we can not keep up with the changes that are being done to the React counterparts of BlazorFluentUI

> My biggest question right now is how I get from web components to something like BlazorFLuentUI if work on this project stops. Where will I get SplitButtons from, what about Models and Panels?

Either keep using BlazorFluentUI in it's current form (again, it is not going away) or start exploring what is possible with Microsoft.Fast.Components.FluentUI. When you find something that is not there, file an issue, contribute and help make it a worthy 'successor'
 
> Is the aim of the FAST web components to supersede what you and other have already achieved in BlazorFluentUI?

The aim is to provide a lightweight set of Blazor component wrappers around Microsoft's official FluentUI Web Components. I can't comment on what kind of level of completeness is being strived for.

**Comment by Cobra86** [comment]
I really like this library, its much better than any other blazor libraries and I'm happy to help with fixing bugs and refactoring. I can't comment about FAST because i never used it. i will try to check out this weekend.
